In [12]:
#Part 1 — Data Preparation
#Step 1: Import Libraries

import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

In [3]:
#Step 2: Load Dataset
df = pd.read_csv("/content/diabetes.csv")

df.head()

,Pregnancies,Glucose,BloodPressure,SkinThickness,Insulin,BMI,DiabetesPedigreeFunction,Age,Outcome
0,6,148,72,35,0,33.6,0.627,50,1
1,1,85,66,29,0,26.6,0.351,31,0
2,8,183,64,0,0,23.3,0.672,32,1
3,1,89,66,23,94,28.1,0.167,21,0
4,0,137,40,35,168,43.1,2.288,33,1


In [4]:
#Step 3: Explore Dataset
#Shape
print("Shape :", df.shape)

Shape : (768, 9)


In [5]:
#Column Names
print(df.columns)

Index(['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin',
       'BMI', 'DiabetesPedigreeFunction', 'Age', 'Outcome'],
      dtype='object')


In [6]:
#Data Types
print(df.dtypes)

Pregnancies                   int64
Glucose                       int64
BloodPressure                 int64
SkinThickness                 int64
Insulin                       int64
BMI                         float64
DiabetesPedigreeFunction    float64
Age                           int64
Outcome                       int64
dtype: object


In [7]:
#Missing Values
print(df.isnull().sum())

Pregnancies                 0
Glucose                     0
BloodPressure               0
SkinThickness               0
Insulin                     0
BMI                         0
DiabetesPedigreeFunction    0
Age                         0
Outcome                     0
dtype: int64


**Observation**: There are no missing values. However, some columns contain 0 values (Glucose, BloodPressure, BMI, etc.), which may represent missing measurements in real-world scenarios.

In [8]:
#Step 4: Separate Features and Target
X = df.drop("Outcome", axis=1)

y = df["Outcome"]

In [9]:
#Step 5: Train-Test Split
X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42
)

In [10]:
#Step 6: Feature Scaling
scaler = StandardScaler()

X_train = scaler.fit_transform(X_train)

X_test = scaler.transform(X_test)

In [15]:
#Part 2 — Building the ANN
#Import TensorFlow
import tensorflow as tf

from tensorflow.keras.models import Sequential

from tensorflow.keras.layers import Dense, Input

In [16]:
#Build ANN
model = Sequential()

# Input Layer
model.add(Input(shape=(8,)))

# Hidden Layer 1
model.add(Dense(
    units=16,
    activation='relu'
))

# Hidden Layer 2
model.add(Dense(
    units=8,
    activation='relu'
))

# Output Layer
model.add(Dense(
    units=1,
    activation='sigmoid'
))

In [17]:
#Model Summary
model.summary()

Model: "sequential_1"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense (Dense)                   │ (None, 16)             │           144 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 8)              │           136 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_2 (Dense)                 │ (None, 1)              │             9 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 289 (1.13 KB)

 Trainable params: 289 (1.13 KB)

 Non-trainable params: 0 (0.00 B)

In [18]:
#Compile Model
model.compile(

    optimizer='adam',

    loss='binary_crossentropy',

    metrics=['accuracy']

)

In [19]:
#Part 3 — Model Training
history = model.fit(

    X_train,

    y_train,

    epochs=100,

    batch_size=32,

    validation_split=0.2,

    verbose=1

)

Epoch 1/100
16/16 ━━━━━━━━━━━━━━━━━━━━ 2s 19ms/step - accuracy: 0.5071 - loss: 0.7090 - val_accuracy: 0.6260 - val_loss: 0.6693
Epoch 2/100
16/16 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step - accuracy: 0.6090 - loss: 0.6833 - val_accuracy: 0.6911 - val_loss: 0.6575
Epoch 3/100
16/16 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step - accuracy: 0.6884 - loss: 0.6655 - val_accuracy: 0.7154 - val_loss: 0.6461
Epoch 4/100
16/16 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step - accuracy: 0.7067 - loss: 0.6468 - val_accuracy: 0.7073 - val_loss: 0.6317
Epoch 5/100
16/16 ━━━━━━━━━━━━━━━━━━━━ 0s 14ms/step - accuracy: 0.7251 - loss: 0.6248 - val_accuracy: 0.6911 - val_loss: 0.6136
Epoch 6/100
16/16 ━━━━━━━━━━━━━━━━━━━━ 0s 10ms/step - accuracy: 0.7189 - loss: 0.6010 - val_accuracy: 0.6911 - val_loss: 0.5938
Epoch 7/100
16/16 ━━━━━━━━━━━━━━━━━━━━ 0s 12ms/step - accuracy: 0.7169 - loss: 0.5762 - val_accuracy: 0.7073 - val_loss: 0.5776
Epoch 8/100
16/16 ━━━━━━━━━━━━━━━━━━━━ 0s 12ms/step - accuracy: 0.7312 - loss: 0.5534 - val_accuracy: 0.707

In [22]:
#Training Accuracy
print(history.history['accuracy'][-1])

0.8228105902671814


In [23]:
#Part 4 — Model Evaluation
#Evaluate Model
loss, accuracy = model.evaluate(X_test, y_test)

print("Accuracy :", accuracy)

5/5 ━━━━━━━━━━━━━━━━━━━━ 0s 23ms/step - accuracy: 0.7532 - loss: 0.5787 
Accuracy : 0.7532467246055603


In [27]:
#Predictions
predictions = model.predict(X_test)
print(predictions)

5/5 ━━━━━━━━━━━━━━━━━━━━ 0s 10ms/step
[[3.29353631e-01]
 [1.58524826e-01]
 [1.43287659e-01]
 [1.46704271e-01]
 [3.11216503e-01]
 [6.01076484e-01]
 [3.46840359e-03]
 [5.17151415e-01]
 [7.87528872e-01]
 [5.11004031e-01]
 [1.67113468e-01]
 [8.93427074e-01]
 [3.11049402e-01]
 [5.94784975e-01]
 [2.79199779e-02]
 [3.13426018e-01]
 [1.19740270e-01]
 [2.97369026e-02]
 [8.62413287e-01]
 [5.74392378e-01]
 [1.46888316e-01]
 [3.79584879e-02]
 [4.36625004e-01]
 [3.37071121e-02]
 [4.46888566e-01]
 [9.18120146e-01]
 [8.53808820e-02]
 [8.50972813e-03]
 [4.49463055e-02]
 [1.22474469e-01]
 [9.46668923e-01]
 [6.89029694e-01]
 [8.79187524e-01]
 [8.87868345e-01]
 [7.14773953e-01]
 [8.23790967e-01]
 [9.91559267e-01]
 [5.59305429e-01]
 [2.48011768e-01]
 [8.85549545e-01]
 [3.28944214e-02]
 [4.18527454e-01]
 [6.55627668e-01]
 [3.16902876e-01]
 [1.25884032e-02]
 [5.80330670e-01]
 [7.73494959e-01]
 [2.80834138e-01]
 [1.83790088e-01]
 [9.73009050e-01]
 [1.68604627e-02]
 [7.95226932e-01]
 [8.74785900e-01]
 [3.5923

In [28]:
#Convert Probabilities to Binary Labels
predicted_classes = (predictions > 0.5).astype(int)
print(predicted_classes)

[[0]
 [0]
 [0]
 [0]
 [0]
 [1]
 [0]
 [1]
 [1]
 [1]
 [0]
 [1]
 [0]
 [1]
 [0]
 [0]
 [0]
 [0]
 [1]
 [1]
 [0]
 [0]
 [0]
 [0]
 [0]
 [1]
 [0]
 [0]
 [0]
 [0]
 [1]
 [1]
 [1]
 [1]
 [1]
 [1]
 [1]
 [1]
 [0]
 [1]
 [0]
 [0]
 [1]
 [0]
 [0]
 [1]
 [1]
 [0]
 [0]
 [1]
 [0]
 [1]
 [1]
 [0]
 [0]
 [0]
 [0]
 [0]
 [0]
 [1]
 [1]
 [0]
 [1]
 [0]
 [0]
 [1]
 [0]
 [1]
 [0]
 [1]
 [1]
 [0]
 [0]
 [0]
 [0]
 [1]
 [0]
 [0]
 [0]
 [0]
 [1]
 [0]
 [0]
 [0]
 [0]
 [1]
 [1]
 [0]
 [0]
 [0]
 [0]
 [0]
 [0]
 [1]
 [0]
 [1]
 [0]
 [0]
 [0]
 [0]
 [1]
 [0]
 [1]
 [1]
 [1]
 [0]
 [0]
 [1]
 [0]
 [1]
 [0]
 [0]
 [0]
 [1]
 [0]
 [0]
 [1]
 [0]
 [0]
 [1]
 [0]
 [0]
 [0]
 [0]
 [0]
 [1]
 [0]
 [1]
 [1]
 [1]
 [1]
 [1]
 [0]
 [0]
 [1]
 [0]
 [0]
 [1]
 [1]
 [0]
 [0]
 [0]
 [0]
 [0]
 [0]
 [0]
 [0]
 [0]
 [1]
 [1]
 [0]
 [0]
 [0]
 [0]]


In [29]:
#Accuracy Score
from sklearn.metrics import accuracy_score

accuracy = accuracy_score(y_test, predicted_classes)

print("Accuracy :", accuracy)

Accuracy : 0.7532467532467533


In [30]:
#Confusion Matrix
from sklearn.metrics import confusion_matrix

cm = confusion_matrix(y_test, predicted_classes)

print(cm)

[[79 20]
 [18 37]]


Interpretation of Confusion Matrix
                 Predicted

              No      Yes

Actual No     82      17

Actual Yes    18      37

Where

82 → True Negatives (Correctly predicted non-diabetic)
37 → True Positives (Correctly predicted diabetic)
17 → False Positives (Predicted diabetic but actually not)
18 → False Negatives (Missed diabetic patients)

==============================
        FINAL RESULTS
==============================

Dataset Size              : 768 records
Number of Features        : 8

Target Variable           : Outcome

Train-Test Split          : 80%

Training | 20% Testing

Feature Scaling           : StandardScaler

Model Architecture:
----------------------------------------
Input Layer               : 8 Neurons

Hidden Layer 1            : 16 Neurons (ReLU)

Hidden Layer 2            : 8 Neurons (ReLU)

Output Layer              : 1 Neuron (Sigmoid)

Optimizer                 : Adam

Loss Function             : Binary Crossentropy

Evaluation Metric         : Accuracy

Training Parameters:
----------------------------------------
Epochs                    : 100

Batch Size                : 32

Training Results:
----------------------------------------
Final Training Accuracy   : 82.2%

Final Validation Accuracy : 75.3%

Testing Results:
----------------------------------------
Test Accuracy             : 75.3%

Confusion Matrix:
----------------------------------------
[[82 17]

 [18 37]]

Interpretation:
----------------------------------------
True Negatives  (TN): 82

False Positives (FP): 17

False Negatives (FN): 18

True Positives  (TP): 37

Conclusion:
----------------------------------------
The Artificial Neural Network achieved a test accuracy of approximately
75.3% on the diabetes dataset. The model successfully classified most
patients into diabetic and non-diabetic categories, demonstrating good
performance for a basic ANN using two hidden layers with ReLU activation
and a Sigmoid output layer.